src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestStream.cs

```diff
@@ -227,5 +249,184 @@ private async Task CopyToAsyncInternal(Stream destination, CancellationToken can
                 }
             }
         }
+
+        private class AsyncPipeReader : IValueTaskSource<int>
```

Comment:
Could this be an internal class instead of copied twice?

Reply:
Have it as an `IAsyncEnumerator<int>` read adapter?

Reply:
Possibly. And then just ignore the result. Duplicating the code does make it easy to fix an issue in one place but not the other.

Reply:
Centralized